### PR TITLE
Update Libundwind to 1.6.2, ABI breakage

### PR DESCRIPTION
**Pull request by mkitti**
See: JuliaLang/julia#43520

ABI breakage:
https://abi-laboratory.pro/index.php?view=timeline&l=libunwind

Arch Linux patch to Julia:
https://github.com/archlinux/svntogit-community/blob/4d9e973e1ed5753966229f40aea2dc7ac77bb889/trunk/julia-libunwind-1.6.patch

### CODE REVIEW

File: L/LibUnwind/LibUnwind@1.5.0/build_tarballs.jl [hunk 1]
```diff
@@ -1,16 +1,12 @@
 using BinaryBuilder
 
 name = "LibUnwind"
-version = v"1.5.0"
-
-# XXX: The git tag for libunwind v1.5.0 is just v1.5 rather than v1.5.0. Check this
-# when updating to the next libunwind version.
-version_short = "1.5"
+version = v"1.6.2"
```
Comment (giordano) [inline L/LibUnwind/LibUnwind@1.5.0/build_tarballs.jl:4]: Did you notice this directory is called `LibUnwind@1.5.0`?

Comment (mkitti) [reply L/LibUnwind/LibUnwind@1.5.0/build_tarballs.jl:4]: oh. no

File: L/LibUnwind/LibUnwind@1.6.2/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+atomic_patch -p0 ${WORKSPACE}/srcdir/patches/libunwind-configure-static-lzma.patch
+atomic_patch -p1 ${WORKSPACE}/srcdir/patches/libunwind-cfa-rsp.patch
+atomic_patch -p1 ${WORKSPACE}/srcdir/patches/libunwind-dwarf-table.patch
+
+CFLAGS="${CFLAGS} -DPI -fPIC -I${prefix}/include"
+./configure \
+    --prefix=${prefix} \
+    --build=${MACHTYPE} \
+    --host=${target} \
+    CFLAGS="${CFLAGS}" \
```
Comment (giordano) [inline L/LibUnwind/LibUnwind@1.6.2/build_tarballs.jl:34]:
```suggestion
```

File: L/LibUnwind/LibUnwind@1.6.2/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    autoreconf -vi
+fi
+
+atomic_patch -p1 ${WORKSPACE}/srcdir/patches/libunwind-prefer-extbl.patch
+atomic_patch -p0 ${WORKSPACE}/srcdir/patches/libunwind-configure-ppc64le.patch
+atomic_patch -p0 ${WORKSPACE}/srcdir/patches/libunwind-configure-static-lzma.patch
+atomic_patch -p1 ${WORKSPACE}/srcdir/patches/libunwind-cfa-rsp.patch
+atomic_patch -p1 ${WORKSPACE}/srcdir/patches/libunwind-dwarf-table.patch
+
+CFLAGS="${CFLAGS} -DPI -fPIC -I${prefix}/include"
```
Comment (giordano) [inline L/LibUnwind/LibUnwind@1.6.2/build_tarballs.jl:29]:
```suggestion
export CFLAGS="-DPI -fPIC -I${includedir}"
```

**Review by giordano — COMMENTED** [review]
It looks like you have to fix the patches

File: L/LibUnwind/LibUnwind@1.6.2/build_tarballs.jl [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+(cd unpacked/liblzma; ar -x ${prefix}/lib/liblzma.a)
+(cd unpacked/libunwind; ar -x ${prefix}/lib/libunwind.a)
+rm -f ${prefix}/lib/libunwind.a
+ar -qc ${prefix}/lib/libunwind.a unpacked/**/*
+"""
+
+# These are the platforms we will build for by default, unless further
+# platforms are passed in on the command line.  libunwind is only used
+# on Linux or FreeBSD (e.g. ELF systems)
+platforms = filter(p -> Sys.islinux(p) || Sys.isfreebsd(p), supported_platforms(;experimental=true))
```
Comment (giordano) [inline L/LibUnwind/LibUnwind@1.6.2/build_tarballs.jl:53]:
```suggestion
platforms = filter(p -> Sys.islinux(p) || Sys.isfreebsd(p), supported_platforms())
```

**Comment by mkitti** [comment]
> It looks like you have to fix the patches

Yes, I'm working through that. Thanks for the help.

**Comment by mkitti** [comment]
Remove libunwind-static-arm.patch, integrated upstream
https://github.com/libunwind/libunwind/commit/1dd0b64fa9d0be66252f7f66e530fffbd59dde86

Remove libunwind-configure-ppc641e.patch
https://github.com/libunwind/libunwind/commit/0af7e7a53480ce8e1cf6cfb4e9fe071c1185ef31

Apply patch via git (updated libunwind-dwarf-table.patch cc: @ararslan )
https://github.com/libunwind/libunwind/commit/8b3832fea8a343ec0888545088c859074aaee20f

File: L/LibUnwind/LibUnwind@1.6.2/build_tarballs.jl [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = [
+    BuildDependency("XZ_jll"),
+    Dependency("Zlib_jll"),
+]
+
+# Build the tarballs.
+build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; julia_compat="1.7")
```
Comment (giordano) [inline L/LibUnwind/LibUnwind@1.6.2/build_tarballs.jl:65]: It looks like `stdatomics.h` requires GCC 4.9: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=58016
```suggestion
build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; preferred_gcc_version=v"5", julia_compat="1.7")
```
Also, `julia_compat="1.7"` is definitely wrong.

File: L/LibUnwind/LibUnwind@1.6.2/build_tarballs.jl [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+# Bash recipe for building across all platforms
+script = raw"""
+cd $WORKSPACE/srcdir/libunwind*/
+
+# We need to massage configure script to convince it to build the shared library
+# for PowerPC.
+if [[ "${target}" == powerpc64le-* ]]; then
+    autoreconf -vi
+fi
+
```
Comment (giordano) [inline L/LibUnwind/LibUnwind@1.6.2/build_tarballs.jl:22]: It looks like we don't need this anymore.

**Comment by mkitti** [comment]
I'm guessing that this is the root our current problems on aarch-linux-musl:

```
[00:47:19] /opt/aarch64-linux-musl/aarch64-linux-musl/sys-root/usr/include/asm/sigcontext.h:28:8: error: redefinition of ‘struct sigcontext’
[00:47:19]  struct sigcontext {
```

This seem to be similar to other issues:
https://bugs.gentoo.org/677098

**Comment by imciner2** [comment]
Julia is now on 1.7.2, so closing this.

**Pull request closed without merging**